### PR TITLE
Update gather_all_tensors to handle tensors of different sizes

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

This was fixed in torchmetrics but not in Lightning
Reference https://github.com/PyTorchLightning/metrics/pull/220

### Does your PR introduce any breaking changes? If yes, please list them.
None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @awaelchli @rohitgr7 @akihironitta

### CODE REVIEW

File: pytorch_lightning/utilities/distributed.py [hunk 1]
```diff
@@ -47,10 +47,11 @@ class group:  # type: ignore
 def gather_all_tensors(result: torch.Tensor, group: Optional[Any] = None) -> List[torch.Tensor]:
     """Function to gather all tensors from several ddp processes onto a list that is broadcasted to all processes.
 
+    Works on tensors that have the same number of dimensions, but where each dimension may differ. In this case
+    tensors are padded, gathered and then trimmed to secure equal workload for all processes.
```
Comment (carmocca) [inline pytorch_lightning/utilities/distributed.py:51]:
```suggestion
    tensors are padded, gathered and then trimmed to secure equal workload for all processes.

```

File: pytorch_lightning/utilities/distributed.py [hunk 2]
```diff
@@ -62,13 +63,35 @@ def gather_all_tensors(result: torch.Tensor, group: Optional[Any] = None) -> Lis
...
-
-    gathered_result = [torch.zeros_like(result) for _ in range(world_size)]
-
-    # sync and broadcast all
     torch.distributed.barrier(group=group)
-    torch.distributed.all_gather(gathered_result, result, group)
 
+    # if the tensor is scalar, things are easy
+    if result.ndim == 0:
+        return _simple_gather_all_tensors(result, group, world_size)
```
Comment (carmocca) [inline pytorch_lightning/utilities/distributed.py:70]: You didn't define this function

File: pytorch_lightning/utilities/distributed.py [hunk 3]
```diff
@@ -62,13 +63,35 @@ def gather_all_tensors(result: torch.Tensor, group: Optional[Any] = None) -> Lis
...
+
+    # 1. Gather sizes of all tensors
+    local_size = torch.tensor(result.shape, device=result.device)
+    local_sizes = [torch.zeros_like(local_size) for _ in range(world_size)]
```
Comment (carmocca) [inline pytorch_lightning/utilities/distributed.py:74]: The tests from the torchmetrics PR should be brought over

File: pytorch_lightning/utilities/distributed.py [hunk 4]
```diff
@@ -47,10 +47,11 @@ class group:  # type: ignore
 def gather_all_tensors(result: torch.Tensor, group: Optional[Any] = None) -> List[torch.Tensor]:
     """Function to gather all tensors from several ddp processes onto a list that is broadcasted to all processes.
 
+    Works on tensors that have the same number of dimensions, but where each dimension may differ. In this case
```
Comment (carmocca) [inline pytorch_lightning/utilities/distributed.py:50]: Mention this improvement in the CHANGELOG

File: tests/utilities/test_distributed.py [hunk 5]
```diff
@@ -42,3 +43,48 @@ def test_collect_states():
...
+    tensor = torch.ones(rank + 1, 2 - rank)
+    result = gather_all_tensors(tensor)
+    assert len(result) == world_size
+    for idx in range(world_size):
+        val = result[idx]
+        assert val.shape == (idx + 1, 2 - idx)
+        assert (val == torch.ones_like(val)).all()
+
+
+@RunIf(min_gpus=2, min_torch="1.10", skip_windows=True)
```
Comment (awaelchli) [inline tests/utilities/test_distributed.py:80]: is this necessary? i.e., is any of the implementation dependent on GPU or would gloo on cpu work too?

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/utilities/distributed.py [hunk 6]
```diff
@@ -62,13 +65,41 @@ def gather_all_tensors(result: torch.Tensor, group: Optional[Any] = None) -> Lis
...
+    max_size = torch.stack(local_sizes).max(dim=0).values
+    all_sizes_equal = all(all(ls == max_size) for ls in local_sizes)
+
+    # 2. If shapes are all the same, then do a simple gather:
+    if all_sizes_equal:
+        return _simple_gather_all_tensors(result, group, world_size)
+
+    # 3. If not, we need to pad each local tensor to maximum size, gather and then truncate
+    pad_dims = []
+    pad_by = (max_size - local_size).detach().cpu()
```
Comment (four4fish) [inline pytorch_lightning/utilities/distributed.py:87]: Why do we want to pad on cpu?

File: pytorch_lightning/utilities/distributed.py [hunk 7]
```diff
@@ -62,13 +65,41 @@ def gather_all_tensors(result: torch.Tensor, group: Optional[Any] = None) -> Lis
...
+    # if the tensor is scalar, things are easy
+    if result.ndim == 0:
+        return _simple_gather_all_tensors(result, group, world_size)
+
+    # 1. Gather sizes of all tensors
+    local_size = torch.tensor(result.shape, device=result.device)
+    local_sizes = [torch.zeros_like(local_size) for _ in range(world_size)]
+    torch.distributed.all_gather(local_sizes, local_size, group=group)
+    max_size = torch.stack(local_sizes).max(dim=0).values
+    all_sizes_equal = all(all(ls == max_size) for ls in local_sizes)
```
Comment (four4fish) [inline pytorch_lightning/utilities/distributed.py:79]: n00b question, is this only used by metrics? Is that possible that user use `gather_all_tensors` in each steps for some other purpose? If yes, this will add one more collective communication for every all_gather, do we want to make this optional?

**Review by four4fish — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/utilities/distributed.py (hunk 7 above)
Comment (DuYicong515) [reply pytorch_lightning/utilities/distributed.py:79]: Have a similar question as well, should size-checking and padding become part of the default in `gather_all_tensors`? 

Maybe we can add as an option which could be turned on/off instead, so that it avoid some of the complexities and overheads if not necessary?

**Review by DuYicong515 — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]
